### PR TITLE
feat: Generate Python client code using GraphQL introspection

### LINKED ISSUE #894: Python client models should be automatically derived from the GraphQL schema
The model classes in the Python client are manually maintained in [_models.py](https://github.com/chanzuckerberg/cryoet-data-portal/blob/6a38f11afe46213e247d204a272d6b49165e097b/client/python/cryoet_data_portal/src/cryoet_data_portal/_models.py).

The class and field names are almost entirely defined in the GraphQL schema, which is [also in this repository](https://github.com/chanzuckerberg/cryoet-data-portal/blob/6a38f11afe46213e247d204a272d6b49165e097b/client/python/cryoet_data_portal/src/cryoet_data_portal/data/schema.graphql).

Maintaining this manually [has created documentation errors in the past](https://github.com/chanzuckerberg/cryoet-data-portal/pull/282), and may lead to bigger issues in the future when the schema changes.

The goal here is to automatically derive these model classes from the GraphQL schema.

### PR DESCRIPTION
**Pull request by andy-sweet**
Closes #894

This adds a custom code generator for the Python client. The mapping between GraphQL types, model types, and fields is somewhat explicit, but doesn't require too much code. The code generation updates the schema and then parses it using de-factor GraphQL python library (which we already depend on). We then use Jinja2 templates to populate the models module.

I [tried some existing packages for generating Python GraphQL clients](https://github.com/chanzuckerberg/cryoet-data-portal/issues/894#issuecomment-2239561088), and my sense was that writing something specific for this client is probably easier than learning how to configure those tools for this particular client. This approach also has the advantage not imposing its own client and query style, which makes this compatible with the existing API by design.

I have done a little manual testing, but testing is mostly through the newly modified or added Python tests. These include checking that missing item relationships are handled gracefully by the client, which is at least a temporary requirement as the ingestion runs and may be a more permanent one.

As the prod GraphQL server has been updated ahead of the Python client, these changes also include the corresponding changes to the models module and GraphQL schema file. As well as the associated test infrastructure changes.

### CODE REVIEW

File: client/python/cryoet_data_portal/src/cryoet_data_portal/_codegen.py
Comment (andy-sweet) [inline client/python/cryoet_data_portal/src/cryoet_data_portal/_codegen.py:39]: Unsure how this value should be parsed into `Annotation.method_links`.

Should this be a `str` (as now) or a `dict`? I still need to look at real live values to understand this.

File: client/python/cryoet_data_portal/tests/test_codegen.py
Comment (andy-sweet) [inline client/python/cryoet_data_portal/tests/test_codegen.py:6]: I still need much more test coverage for code generator. I think just a few more tests for each of the public functions might be good enough.

File: client/python/cryoet_data_portal/tests/test_get_client.py
Comment (andy-sweet) [inline client/python/cryoet_data_portal/tests/test_get_client.py:9]: I modified these tests to account for depositions and deposition authors in a way that seems consistent with what came before.

I think these could generally be more meaningful, but this may not be the right time to think too deeply. Open to ideas though.

**Comment by andy-sweet** [comment]
A few things are blocking this from be reviewable and thus mergeable.

1. I should increase code coverage by covering the code generator. I think we should be able to reuse the existing test infrastructure to do this in a fairly meaningful way without needing to make new complex mocks or stubs.

2. It seems like the `deposition_id` can be `None` in production. Assuming this is not temporary, I need to add some logic to the relationship field types to handle that case. We should also have at least one test to cover this too.

3. I need to determine what the type and value of `Annotation.method_links` should be (i.e. `str` vs. `dict`) and if that needs any special handling for the `dict`.

Ideally this PR would be much smaller, but I'm not sure how easy that is given that this mostly generates code/data and quite a few things have changed on prod.

File: client/python/cryoet_data_portal/src/cryoet_data_portal/_models.py
Comment (andy-sweet) [inline client/python/cryoet_data_portal/src/cryoet_data_portal/_models.py:751]: Note that `dataset` is the name in the schema and not `datasets` as we might hope here. I assume it's that to avoid a name clash, but I'm not sure. Not ideal, but we can probably live with it. We could also provide a `datasets` property alias in a Jinja template to improve the situation.

File: client/python/cryoet_data_portal/src/cryoet_data_portal/_codegen.py
Comment (manasaV3) [reply client/python/cryoet_data_portal/src/cryoet_data_portal/_codegen.py:39]: This would be a `dict`, as it would be a JsonField. :)

Comment (andy-sweet) [reply client/python/cryoet_data_portal/src/cryoet_data_portal/_codegen.py]: I removed it for now (so it's ignored) since @jgadling was unsure if this should be live and public. My plan is to add it in a follow-up PR since this one is more than large enough!

**Comment by andy-sweet** [comment]
This is ready for review now.

The PR is *much* larger than I'd like (though almost all the delta is auto-generated in `_models.py`, `schema.graphql`, and `up.sql`), but I am unsure how to split this up without setting up new fakes/stubs that may take a significant amount of time and probably are not worth it in the long term.

There are two practical issues blocking this from being merged.

1) The tests I added exacerbate the flakiness described in #942, so we should almost certainly fix that before merging this.
2) One of my fancy tests that programmatically imports a temp module seems to break codecov. I should either change the test or find the right way to make codecov ignore that module.

File: client/python/cryoet_data_portal/Makefile
Comment (jgadling) [inline client/python/cryoet_data_portal/Makefile:30]: Should we just use ruff? afaik newer versions of ruff make black kindof unnecessary?

File: client/python/cryoet_data_portal/src/cryoet_data_portal/_models.py
Comment (jgadling) [inline client/python/cryoet_data_portal/src/cryoet_data_portal/_models.py:76]: Switching to strings instead of passing classes around is nice! I forgot that the Relationship class can handle this!

File: client/python/cryoet_data_portal/Makefile
Comment (andy-sweet) [reply client/python/cryoet_data_portal/Makefile:30]: I think @kne42 originally added these about a year ago, so checking if we can consolidate to just ruff since it may have improved rule coverage since then.

Comment (jgadling) [reply client/python/cryoet_data_portal/Makefile:30]: I'm in favor of just using one tool unless there's something major we'd be missing out on. @manasaV3 ?

**Review by jgadling — APPROVED** [review]
This looks great! I think we'll need to update the docs templates to add depositions / deposition authors when we get a chance but that can be handled in a later PR.

**Comment by andy-sweet** [comment]
> This looks great! I think we'll need to update the docs templates to add depositions / deposition authors when we get a chance but that can be handled in a later PR.

If there's no existing issue I'll add an one after merging this to make sure that doesn't get lost.

**Comment by andy-sweet** [comment]
> One of my fancy tests that programmatically imports a temp module seems to break codecov. I should either change the test or find the right way to make codecov ignore that module.

I added a `.coveragerc` file to the python client directory and explicitly ignored the temp module I create. That's possibly sensitive to changes, but I don't want to ignore things that should not be ignored.

File: client/python/cryoet_data_portal/Makefile
Comment (manasaV3) [reply client/python/cryoet_data_portal/Makefile:30]: That sounds good to me! :)

**Review by manasaV3 — APPROVED** [review]

**Pull request merged**
